Desktop_Interface/isobuffer.cpp

```diff
@@ -2,391 +2,364 @@
-
-    int front = back - 1 - delaySamples;
-    if (front < 0) front = 0;
-    int idx, subIdx;
-    if(readData!=NULL) free(readData);
-    readData = (short *) calloc(numSamples, sizeof(short));
-
-    if(singleBit){
-        for (int i=0; i<numSamples;i++){
-            if (timeBetweenSamples > (double) front){
-                accumulatedDelay -= (double) front;
-                front = bufferEnd;
-            }
-
-            idx = (int) floor(((double) front - accumulatedDelay));
-            subIdx = (int) floor(8*(((double) front - accumulatedDelay) - floor(((double) front - accumulatedDelay))));
-
-            //qDebug() << "subIdx = " << subIdx;
-
-            if (idx < 0){
-                accumulatedDelay--;
-                accumulatedDelay -= (double) front;
-                front = bufferEnd;
-                idx = (int) round(((double) front - accumulatedDelay));
-            }
-            readData[i] = buffer[idx] & (1 << subIdx);
-
-            accumulatedDelay += timeBetweenSamples;
-        }
-    }else{
-        for (int i=0; i<numSamples;i++){
-            if (timeBetweenSamples > (double) front){
-                accumulatedDelay -= (double) front;
-                front = bufferEnd;
-            }
-
-            idx = (int) round(((double) front - accumulatedDelay));
-            if (idx < 0){
-                accumulatedDelay--;
-                accumulatedDelay -= (double) front;
-                front = bufferEnd;
-                idx = (int) round(((double) front - accumulatedDelay));
-            }
-            readData[i] = buffer[idx];
-
-            accumulatedDelay += timeBetweenSamples;
-        }
-    }
-
-    return readData;
+	/* Refactor Note:
+	 *
+	 * Refactoring this function took a few passes were i made some assumptions:
+	 *  - round() should be replaced by floor() where it was used
+	 *  - int(floor(x)) and int(x) are equivalent (since we are always positive)
+	 *  - free(NULL) is a no-op. This is mandated by the C standard, and virtually all
+	 * implementations comply. A few known exceptions are:
+	 *    - PalmOS
+	 *    - 3BSD
+	 *    - UNIX 7
+	 *   I do not know of any non-compliant somewhat modern implementations.
+	 *
+	 * The expected behavior is to cycle backwards over the buffer with a stride of
+	 * timeBetweenSamples steps, and insert the touched elements into readBuffer.
+	 *
+	 * ~Sebastian Mestre
+	 */
+	const double timeBetweenSamples = sampleWindow * samplesPerSecond / numSamples;
+	const int delaySamples = delayOffset * samplesPerSecond;
+
+	free(readData);
+	readData = (short*) calloc(numSamples, sizeof(short));
+
+	double itr = delaySamples + 1;
+	for (int i = 0; i < numSamples; i++)
+	{
+		while (itr > insertedCount)
+			itr -= insertedCount;
+
+		readData[i] = bufferAt(int(itr));
+
+		if (singleBit)
+		{
+			int subIdx = 8*(-itr-floor(-itr));
+			readData[i] &= (1 << subIdx);
+		}
+
+		itr += timeBetweenSamples;
+	}
+
+	return readData;
```

Comment:
I'm not sure that returning a plain pointer is the best way to go here, as it means that the caller must manually free it (and who owns the pointer is not always clear!).  Since we have C++11, std::unique_ptr and std::move could solve this problem.

Reply:
Ah yes. I'll take some modifying of caller code though, so I'll put it off till the next PR